### PR TITLE
Case 7120, Case 14963: Procedural material entities on primitives and meshes

### PR DESCRIPTION
**Pull request by SamGondelman**
http://roadmap.highfidelity.com/open-source-project-proposals/p/procedural-materials-on-models-and-avatars
https://highfidelity.manuscript.com/f/cases/7120/Ability-for-a-user-to-create-import-custom-shaders-as-an-entity-material-type
https://highfidelity.fogbugz.com/f/cases/14963/Add-support-for-shaders-on-meshes-and-avatars

Test plan:
- A new option has been added in Settings -> Graphics and at the bottom of Developer -> Render, "Use Procedural Materials on Meshes".  It will default to disabled.
- Run [this](https://gist.githubusercontent.com/SamGondelman/df6f6c63bc8297a7d4a2dd439b72d8fb/raw/d7eb2a0e565b9e5683ab7981bca8a458436c57d3/ProceduralMaterial.js).  You'll see this:
hifi-snap-by-SamGondelman-on-2019-05-16_20-14-34
- When you press "u", the material entity (the small sphere) will be applied to the following in order and will then repeat: the sphere, the woody, the tablet, your avatar, nothing
- You'll see the shader on the sphere:
hifi-snap-by-SamGondelman-on-2019-05-16_20-14-49
- But you won't see the material on the mannequin, tablet, or yourself (it will make the targeted meshes disappear):
hifi-snap-by-SamGondelman-on-2019-05-16_20-15-23
- When you enable "Use Procedural Materials on Meshes", you will start to see the material on those things:
hifi-snap-by-SamGondelman-on-2019-05-16_20-15-37
hifi-snap-by-SamGondelman-on-2019-05-16_20-15-45
hifi-snap-by-SamGondelman-on-2019-05-16_20-15-53
- The setting will persist between sessions.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19527/HighFidelity-Beta-PR15577-25c2330.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19527/HighFidelity-Beta-Interface-PR15577-25c2330.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19527/HighFidelity-Beta-PR15577-25c2330.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19527/HighFidelity-Beta-Interface-PR15577-25c2330.exe

File: assignment-client/CMakeLists.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@ link_hifi_libraries(
   controllers physics plugins midi image
   material-networking model-networking ktx shaders
 )
+include_hifi_library_headers(procedural)
```
Comment (SamGondelman) [inline assignment-client/CMakeLists.txt]: this PR includes some amount of dependency finagling

File: libraries/avatars/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,3 @@
 set(TARGET_NAME avatars)
 setup_hifi_library(Network Script)
-include_hifi_library_headers(gpu)
-link_hifi_libraries(shared networking graphics)
```
Comment (SamGondelman) [inline libraries/avatars/CMakeLists.txt]: but also some amount of dependency _de_-finagling

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 3]
```diff
@@ -1325,7 +1325,7 @@ void ModelEntityRenderer::doRenderUpdateSynchronousTyped(const ScenePointer& sce
 
     // Check for addition
     if (_hasModel && !model) {
-        model = std::make_shared<Model>(nullptr, entity.get());
+        model = std::make_shared<Model>(nullptr, entity.get(), _created);
```
Comment (SamGondelman) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: Models have a _created timestamp now that they pass to their MeshPartPayloads.  This is the only place it's set, everywhere else there's no sense of the creation time

File: libraries/gpu/src/gpu/Shader.h [hunk 4]
```diff
@@ -118,9 +118,6 @@ class Shader {
     Shader(Type type, const Source& source, bool dynamic);
     Shader(Type type, const Pointer& vertex, const Pointer& geometry, const Pointer& pixel);
 
-    Shader(const Shader& shader);             // deep copy of the sysmem shader
-    Shader& operator=(const Shader& shader);  // deep copy of the sysmem texture
```
Comment (SamGondelman) [inline libraries/gpu/src/gpu/Shader.h]: never defined

File: libraries/graphics-scripting/src/graphics-scripting/GraphicsScriptingInterface.cpp [hunk 5]
```diff
@@ -372,120 +372,125 @@ namespace scriptable {
...
 
-        if (hasPropertyFallthroughs && material.propertyFallthroughs.at(graphics::MaterialKey::GLOSSY_VAL_BIT)) {
-            obj.setProperty("roughness", FALLTHROUGH);
-        } else if (material.key.isGlossy()) {
-            obj.setProperty("roughness", material.roughness);
+        if (hasPropertyFallthroughs && material.propertyFallthroughs.at(graphics::MaterialKey::ALBEDO_VAL_BIT)) {
+            obj.setProperty("albedo", FALLTHROUGH);
+        } else if (material.key.isAlbedo()) {
+            obj.setProperty("albedo", vec3ColorToScriptValue(engine, material.albedo));
         }
```
Comment (SamGondelman) [inline libraries/graphics-scripting/src/graphics-scripting/GraphicsScriptingInterface.cpp]: We set opacity and albedo for both material models.  Everything below here is just moved inside an `if`, and then the `else` handles `procedural`

File: libraries/procedural/src/procedural/Procedural.cpp [hunk 6]
```diff
@@ -410,7 +430,7 @@ void Procedural::setupUniforms() {
         }
 
         _standardInputsBuffer->setSubData(0, _standardInputs);
-        batch.setUniformBuffer(0, _standardInputsBuffer, 0, sizeof(StandardInputs));
+        batch.setUniformBuffer(procedural::slot::buffer::Inputs, _standardInputsBuffer, 0, sizeof(StandardInputs));
```
Comment (SamGondelman) [inline libraries/procedural/src/procedural/Procedural.cpp]: We were using slot 0 for both this and GRAPHICS_BUFFER_SKINNING, so as you'll see below, I changed this.

File: libraries/procedural/src/procedural/ShaderConstants.h [hunk 7]
```diff
@@ -14,17 +14,24 @@
...
 
+#define PROCEDURAL_BUFFER_INPUTS 7
+
 #define PROCEDURAL_UNIFORM_CUSTOM 220
 
-#define PROCEDURAL_TEXTURE_CHANNEL0 0
-#define PROCEDURAL_TEXTURE_CHANNEL1 1
-#define PROCEDURAL_TEXTURE_CHANNEL2 2
-#define PROCEDURAL_TEXTURE_CHANNEL3 3
+#define PROCEDURAL_TEXTURE_CHANNEL0 2
```
Comment (SamGondelman) [inline libraries/procedural/src/procedural/ShaderConstants.h]: I bumped these so they don't collide with any other textures.

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 8]
```diff
@@ -432,9 +463,23 @@ void ModelMeshPartPayload::render(RenderArgs* args) {
         batch.setDrawcallUniform(drawcallInfo);
     }
 
-    // apply material properties
-    if (RenderPipelines::bindMaterials(_drawMaterials, batch, args->_renderMode, args->_enableTexturing)) {
-        args->_details._materialSwitches++;
+    if (!_drawMaterials.empty() && _drawMaterials.top().material && _drawMaterials.top().material->isProcedural() &&
+            _drawMaterials.top().material->isReady()) {
+        if (!enableMeshShaders) {
```
Comment (SamGondelman) [inline libraries/render-utils/src/MeshPartPayload.cpp]: This is where we disable (don't render) shaders on meshes if they're turned off.

File: libraries/render-utils/src/deformed_simple.slv [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+<@include gpu/Inputs.slh@>
+<@include gpu/Color.slh@>
+<@include gpu/Transform.slh@>
+<$declareStandardTransform()$>
+
+<@include MeshDeformer.slh@>
+<$declareMeshDeformer(1, _SCRIBE_NULL, 1, _SCRIBE_NULL, 1)$>
```
Comment (SamGondelman) [inline libraries/render-utils/src/deformed_simple.slv]: These two slv's are straightforward, the only difference is the second to last argument here, which is whether or not we're using DQ skinning.

File: libraries/script-engine/src/SceneScriptingInterface.cpp [hunk 10]
```diff
@@ -11,129 +11,6 @@
 
 #include "SceneScriptingInterface.h"
 
-#include <procedural/ProceduralSkybox.h>
```
Comment (SamGondelman) [inline libraries/script-engine/src/SceneScriptingInterface.cpp]: This stuff is unused and a dependency headache so I've removed it.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19528/HighFidelity-Beta-PR15577-f4f687c.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19528/HighFidelity-Beta-Interface-PR15577-f4f687c.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19528/HighFidelity-Beta-PR15577-f4f687c.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19528/HighFidelity-Quest-Beta-PR15577-f4f687c.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19528/HighFidelity-Beta-PR15577-f4f687c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19528/HighFidelity-Beta-Interface-PR15577-f4f687c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19565/HighFidelity-Beta-PR15577-a7842fe.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19565/HighFidelity-Beta-Interface-PR15577-a7842fe.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19565/HighFidelity-Beta-PR15577-a7842fe.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19565/HighFidelity-Quest-Beta-PR15577-a7842fe.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19565/HighFidelity-Beta-PR15577-a7842fe.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19565/HighFidelity-Beta-Interface-PR15577-a7842fe.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19615/HighFidelity-Beta-PR15577-3d05a95.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19615/HighFidelity-Quest-Beta-PR15577-3d05a95.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19615/HighFidelity-Beta-PR15577-3d05a95.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19615/HighFidelity-Beta-Interface-PR15577-3d05a95.exe

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19624/HighFidelity-Beta-PR15577-3d05a95.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19624/HighFidelity-Quest-Beta-PR15577-3d05a95.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19624/HighFidelity-Beta-PR15577-3d05a95.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19624/HighFidelity-Beta-Interface-PR15577-3d05a95.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19624/HighFidelity-Beta-PR15577-3d05a95.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19624/HighFidelity-Beta-Interface-PR15577-3d05a95.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19753/HighFidelity-Beta-PR15577-16a56dc.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19753/HighFidelity-Beta-Interface-PR15577-16a56dc.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19753/HighFidelity-Beta-PR15577-16a56dc.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19753/HighFidelity-Quest-Beta-PR15577-16a56dc.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19753/HighFidelity-Beta-PR15577-16a56dc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19753/HighFidelity-Beta-Interface-PR15577-16a56dc.exe

File: libraries/render-utils/src/deformed_simple.slv (hunk 9 above)
Comment (SamGondelman) [reply libraries/render-utils/src/deformed_simple.slv]: If we were smart we would merge https://github.com/highfidelity/hifi/pull/15671 first to make this simpler, but...we aren't smart.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19865/HighFidelity-Beta-PR15577-37feb7f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19865/HighFidelity-Beta-Interface-PR15577-37feb7f.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19865/HighFidelity-Beta-PR15577-37feb7f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19865/HighFidelity-Quest-Beta-PR15577-37feb7f.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19865/HighFidelity-Beta-PR15577-37feb7f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19865/HighFidelity-Beta-Interface-PR15577-37feb7f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20101/HighFidelity-Beta-PR15577-e4847cc.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20101/HighFidelity-Beta-Interface-PR15577-e4847cc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20101/HighFidelity-Beta-PR15577-e4847cc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20101/HighFidelity-Beta-Interface-PR15577-e4847cc.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20101/HighFidelity-Beta-PR15577-e4847cc.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20101/HighFidelity-Quest-Beta-PR15577-e4847cc.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20152/HighFidelity-Beta-PR15577-9c1b139.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20152/HighFidelity-Beta-Interface-PR15577-9c1b139.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20152/HighFidelity-Beta-PR15577-9c1b139.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20152/HighFidelity-Beta-Interface-PR15577-9c1b139.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20152/HighFidelity-Beta-PR15577-9c1b139.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20152/HighFidelity-Quest-Beta-PR15577-9c1b139.apk)

File: libraries/render-utils/src/deformed_simple.slv (hunk 9 above)
Comment (SamGondelman) [reply libraries/render-utils/src/deformed_simple.slv]: we are so smart

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20364/HighFidelity-Beta-PR15577-6060077.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20364/HighFidelity-Quest-Beta-PR15577-6060077.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20364/HighFidelity-Beta-PR15577-6060077.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20364/HighFidelity-Beta-Interface-PR15577-6060077.dmg

**Comment by SamGondelman** [comment]
build this please

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20368/HighFidelity-Beta-PR15577-6060077.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20368/HighFidelity-Quest-Beta-PR15577-6060077.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20368/HighFidelity-Beta-PR15577-6060077.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20368/HighFidelity-Beta-Interface-PR15577-6060077.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20368/HighFidelity-Beta-PR15577-6060077.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20368/HighFidelity-Beta-Interface-PR15577-6060077.exe

**Comment by HifiExperiments** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20778/HighFidelity-Beta-PR15577-6060077.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20778/HighFidelity-Quest-Beta-PR15577-6060077.apk)

**Pull request closed without merging**